processor/deltatocumulativeprocessor/benchmark_test.go

```diff
@@ -148,14 +151,95 @@ func BenchmarkProcessor(gb *testing.B) {
 				md.CopyTo(req)
 
 				b.StartTimer()
-				err := st.proc.ConsumeMetrics(ctx, req)
+				err := proc.ConsumeMetrics(ctx, req)
 				b.StopTimer()
 				require.NoError(b, err)
 			}
 
 			// verify all dps are processed without error
 			b.StopTimer()
-			require.Equal(b, b.N*metrics*streams, st.sink.DataPointCount())
+			require.Equal(b, b.N*metrics*streams, sink.DataPointCount())
 		})
 	}
 }
+
+func Benchmark(b *testing.B) {
+	const K = 64               // number of metrics
+	const S = 3                // number of streams
+	P := runtime.GOMAXPROCS(0) // number of routines
+	if P > K {
+		b.Fatal("increase K")
+	}
+
+	var id atomic.Int64
+	id.Store(-1)
+
+	var init sync.WaitGroup
+	init.Add(P)
+	wait := make(chan struct{})
+
+	start := time.Now()
+
+	sink := new(CountingSink)
+	proc, _ := setup(b, nil, sink)
+
+	b.ReportAllocs()
+	go func() {
+		init.Wait()
+		b.ResetTimer()
+		close(wait)
+	}()
+
+	k := K / P
+	b.RunParallel(func(pb *testing.PB) {
+		md := pmetric.NewMetrics()
+		ms := make([]pmetric.Metric, k)
+		for i := range ms {
+			m := md.ResourceMetrics().AppendEmpty().ScopeMetrics().AppendEmpty().Metrics().AppendEmpty()
+			mid := id.Add(1)
+			m.SetName(fmt.Sprintf("metric-%d", mid))
+			sum := m.SetEmptySum()
+			for s := range S {
+				dp := sum.DataPoints().AppendEmpty()
+				dp.Attributes().PutInt("s", int64(s))
+				dp.SetIntValue(rand.Int64N(100))
+				dp.SetStartTimestamp(pcommon.NewTimestampFromTime(start))
+			}
+			ms[i] = m
+		}
+		ctx := context.Background()
+		init.Done()
+
+		<-wait
+		for n := 0; pb.Next(); n++ {
+			for _, m := range ms {
+				m.Sum().SetAggregationTemporality(pmetric.AggregationTemporalityDelta)
+				dps := m.Sum().DataPoints()
+				for i := range dps.Len() {
+					dp := dps.At(i)
+					dp.SetTimestamp(dp.Timestamp() + pcommon.Timestamp(time.Minute.Nanoseconds()))
+				}
+			}
+			if err := proc.ConsumeMetrics(ctx, md); err != nil {
+				b.Fatal(err)
+			}
```

Comment:
Shouldn't we put StartTimer and StopTimer around this? This is what we actually want to benchmark, right?

Reply:
Ok, I just read the doc string for `b.RunParallel`:

> It should not use the [B.StartTimer], [B.StopTimer], or [B.ResetTimer] functions, because they have global effect. It should also not call [B.Run].

I understand why you didn't do it now, but is this the approach we want to follow then? Would it make sense if we ran macrobenchmarks instead?

Reply:
`StartTimer` and `StopTimer` trigger use `runtime.ReadMemStats` which is horribly expensive (it runs a GC and halts everything to record), which distorts benchmarks. 

By doing setup initially and then once the init `WaitGroup` is done resetting the timer avoids that